systemctl: Add timer activation to status

keur opened the pull request:
For all units that aren't timers, if it is activated by a timer unit,
print the timer in systemctl status. For timer units, print the unit
that is fired when the timer activates.

boucman commented:
would it be possible to report other event-source units (.path .socket) ?

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4253,7 +4257,24 @@ static void print_status_info(
+                STRV_FOREACH(trigger, i->triggered_by) {
+                        if (endswith(*trigger, ".timer")) {
+                                if (!found) {
+                                        printf("Activator: Timers\n");
```
We say "Trigger" in .timer unit. For symmetry, I think we should say "TriggeredBy" here. We don't use the term "activator" anywhere else.

If the triggering unit is not active, it will still be shown here. We need to indicate this somehow, maybe even by showing a color dot to indicate the unit status. (There is an assymetry here: when we are showing a .timer unit, and we say 'Triggers: foo.service', it doesn't matter that much whether foo.service is active or not. It will be started by trigger when the time comes either way. But when we are showing the .service unit, if foo.trigger is inactive, no triggering will happen.)

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4253,7 +4257,24 @@ static void print_status_info(
+                bool found = false;
+
+                STRV_FOREACH(trigger, i->triggered_by) {
+                        if (endswith(*trigger, ".timer")) {
```
I have to agree with @boucman: limiting this to timers seems artificial. Maybe just removing this `if` here would be enough?

keur commented:
Fixed and force-pushed. I'm kinda unsure I like how the output looks, so if you have suggestions let me know.

cc @keszybz

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4268,9 +4294,15 @@ static void print_status_info(
+                                char **trigger;
+
+                                STRV_FOREACH(trigger, i->triggers)
+                                        printf("           %s %s\n", special_glyph(SPECIAL_GLYPH_TREE_RIGHT), *trigger);
```
Hmm, I'm not sure if this is clear enough:
```console
$ build/systemctl status unbound-anchor.timer
● unbound-anchor.timer - daily update of the root trust anchor for DNSSEC
   Loaded: loaded (/usr/lib/systemd/system/unbound-anchor.timer; enabled; vendor preset: enabled)
   Active: active (waiting) since Mon 2019-09-16 10:02:49 CEST; 1s ago
  Trigger: Tue 2019-09-17 00:00:00 CEST; 13h left
           └─ unbound-anchor.service
     Docs: man:unbound-anchor(8)
```
Both the trigger time and the triggered units appear under the same heading. Wouldn't it be better to have a separate "Triggers: " section for the triggered units? Also, units that are triggered could be listed even if the triggering unit is stopped.

keszybz replied:
@keur another update?

keszybz commented:
@keur, are you still working on this?

keur commented:
Sorry for the delay. Got caught up with other projects. Changes made.

Also, I opted not to use the code at 504 since it uses unit load status, and instead made a function for the circle code at the beginning of `print_status_info`, since that uses `active_status`.

cc @keszybz

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4303,6 +4333,26 @@ static void print_status_info(
         }
 
+        if (!strv_isempty(i->triggers)) {
+            char **trigger;
```
indentation is off. please stick to multiples of 8ch, see CODING_STYLE

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4303,6 +4333,26 @@ static void print_status_info(
+                    r = get_state_one_unit(bus, *trigger, &state);
+                    if (r < 0) {
+                            continue;
+                    }
```
no {} around single line if blocks

That said, I'd just ignore the error here. i.e. continue and show as is

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4282,7 +4290,29 @@ static void print_status_info(
+                        r = get_state_one_unit(bus, *trigger, &state);
+                        if (r < 0) {
+                                continue;
+                        }
```
no {} around single-line if blocks. As mentioned further down, i'd just ignore the return code here though

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4282,7 +4290,29 @@ static void print_status_info(
+                char **trigger;
+                bool found = false;
+                UnitActiveState state;
+                off = ansi_normal();
```
why this line?

keur commented:
Changes made.

cc @poettering

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4282,7 +4290,25 @@ static void print_status_info(
+
+        if (!is_timer && !strv_isempty(i->triggered_by)) {
+                char **trigger;
+                UnitActiveState state;
```
move this one to the inner loop? Also, please initialize to _UNIT_ACTIVE_STATE_INVALID, so that in the event get_state_one_unit() fails it's still properly initialized to something. And unit_active_state_to_string() then turns that to NULL, which al makes system.

poettering commented on src/systemctl/systemctl.c:
```diff
@@ -4303,6 +4329,22 @@ static void print_status_info(
 
+        if (!strv_isempty(i->triggers)) {
+                char **trigger;
+                UnitActiveState state;
```
ditto

keszybz commented on src/systemctl/systemctl.c:
```diff
@@ -4282,7 +4290,25 @@ static void print_status_info(
+                UnitActiveState state;
+                bool found = false;
+
+                STRV_FOREACH(trigger, i->triggered_by) {
```
Maybe write this a bit different:
```C
if (!strv_isempty(i->triggered_by))
              printf("   TriggeredBy:\n");
STRV_FOREACH(trigger, i->triggered_by) {
              ...
}
```

keur commented:
Changes made

cc @keszybz @poettering

keszybz commented:
The output looks nice, but there was a minor memleak (the two new strvs need to be freed in `unit_status_info_free()`). I force-pushed into your branch with this added. I also added another commit on top, to simplify the `STRV_FOREACH()` loop. It makes no functional change, but shorter is nicer. Also, I removed the check that preventing showing TriggeredBy for timers. We should just print the data that we receive over dbus, interpreting is as little as possible.

keur commented:
On Tue Nov 5, 2019 at 1:08 PM Zbigniew Jędrzejewski-Szmek wrote:
> The output looks nice, but there was a minor memleak (the two new strvs need to be freed in `unit_status_info_free()`). I force-pushed into your branch with this added. I also added another commit on top, to simplify the `STRV_FOREACH()` loop. It makes no functional change, but shorter is nicer. Also, I removed the check that preventing showing TriggeredBy for timers. We should just print the data that we receive over dbus, interpreting is as little as possible.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/systemd/systemd/pull/13554#issuecomment-550020510

Sounds good. Thanks!
